File: testSuite/test-output-tree-contiguousity.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;

# Run models that test that merger trees are output in contiguous blocks.
# Andrew Benson (14-July-2021)

# Make output directory.
system("mkdir -p outputs/");

# Run the model.
system("cd ..; ./Galacticus.exe testSuite/parameters/outputTreeContiguosity.xml");
unless ( $? == 0 ) {
    print "FAIL: output tree contiguosity model failed to run\n";
    exit;
}

# Check for contiguous output.
my $model   = new PDL::IO::HDF5("outputs/outputTreeContiguosity.hdf5");
my $outputs = $model->group('Outputs');
my $success = 1;
foreach my $outputName ( $outputs->groups() ) {
    my $output          = $outputs ->group  ($outputName           )       ;
    my $nodeData        = $output  ->group  ('nodeData'            )       ;
    my $treeIndex       = $output  ->dataset('mergerTreeIndex'     )->get();
    my $treeStart       = $output  ->dataset('mergerTreeStartIndex')->get();
    my $treeCount       = $output  ->dataset('mergerTreeCount'     )->get();
    my $treeIndexByNode = $nodeData->dataset('mergerTreeIndex'     )->get();
    for(my $i=0;$i<nelem($treeIndex);++$i) {
	my $indexTarget =               $treeIndex->(($i));
	my $indexStart  =               $treeStart->(($i));
	my $indexEnd    = $indexStart-1+$treeCount->(($i));
	$success = 0
	    unless ( all($treeIndexByNode->($indexStart:$indexEnd) == $indexTarget) );
    }
}
my $status = $success ? "SUCCESS" : "FAILED";
print $status.": merger tree output contiguosity\n";
exit;

File: testSuite/test-output-tree-contiguousity.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Run models that test that merger trees are output in contiguous blocks.
# Andrew Benson (ported to Python)

# Make output directory.
subprocess.run("mkdir -p outputs/", shell=True)

# Run the model.
status = subprocess.run("cd ..; ./Galacticus.exe testSuite/parameters/outputTreeContiguosity.xml", shell=True)
if status.returncode != 0:
    print("FAIL: output tree contiguosity model failed to run")
    sys.exit(0)

# Check for contiguous output.
success = True
with h5py.File("outputs/outputTreeContiguosity.hdf5", "r") as model:
    outputs = model["Outputs"]
    for outputName in outputs.keys():
        output          = outputs[outputName]
        nodeData        = output["nodeData"]
        treeIndex       = output["mergerTreeIndex"][:]
        treeStart       = output["mergerTreeStartIndex"][:]
        treeCount       = output["mergerTreeCount"][:]
        treeIndexByNode = nodeData["mergerTreeIndex"][:]
        for i in range(len(treeIndex)):
            indexTarget = treeIndex[i]
            indexStart  = treeStart[i]
            indexEnd    = indexStart + treeCount[i]
            if not np.all(treeIndexByNode[indexStart:indexEnd] == indexTarget):
                success = False
                break

status_str = "SUCCESS" if success else "FAILED"
print(f"{status_str}: merger tree output contiguosity")
